libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1037 @@
+                                                  (GDestroyNotify) word_quote_proxy_bundle_free);
+
+  for (i = 0; i < zipped_word_quote_proxies->len; ++i)
+    {
+      WordQuoteProxyBundle *bundle = g_ptr_array_index (zipped_word_quote_proxies, i);
+      append_discovery_feed_word_quote_from_proxies (bundle->word_proxy,
+                                                     bundle->quote_proxy,
+                                                     cancellable,
+                                                     individual_task_result_completed,
+                                                     individual_task_result_closure_new (all_tasks_closure));
+    }
+}
+
+static void
+received_all_unordered_card_array_results_from_queries (GObject      *source,
+                                                        GAsyncResult *result,
+                                                        gpointer      user_data)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GPtrArray) results = g_task_propagate_pointer (G_TASK (result),
+                                                           &local_error);
+  g_autoptr(GPtrArray) result_arrays = g_ptr_array_new_with_free_func ((GDestroyNotify) g_ptr_array_unref);
+  g_autoptr(GPtrArray) flat_mapped_array = NULL;
+  g_autoptr(GTask) task = user_data;
+  guint i = 0;
+
+  /* This basically shouldn't happen, but handle it anyway */
+  if (results == NULL)
+    {
+      g_message ("Error getting all unordered card results: %s", local_error->message);
+      g_task_return_error (task, local_error);
+      return;
+    }
+
+  /* Go through each of the results, complain about the ones that failed
+   * but keep the ones that yielded some models. We'll flat-map them all
+   * together later */
+  for (i = 0; i < results->len; ++i)
+    {
+      g_autoptr(GPtrArray) result_array = g_task_propagate_pointer (g_ptr_array_index (results, i),
+                                                                    &local_error);
+
+      if (result_array == NULL)
+        {
+          g_message ("Query failed: %s", local_error->message);
+          g_clear_error (&local_error);
+          continue;
+        }
+
+      g_ptr_array_add (result_arrays, g_steal_pointer (&result_array));
+    }
+
+  flat_mapped_array = flat_map_ptr_arrays (result_arrays,
+                                           ref_object_copy_func,
+                                           NULL,
+                                           g_object_unref);
+  g_task_return_pointer (task,
+                         g_steal_pointer (&flat_mapped_array),
+                         (GDestroyNotify) g_ptr_array_unref);
+}
+
+/**
+ * eos_discovery_feed_unordered_results_from_queries_finish:
+ * @result: A #GAsyncResult
+ * @error: A #GError
+ *
+ * Complete the call to eos_discovery_feed_unordered_results_from_queries.
+ *
+ * Returns: (transfer none) (element-type EosDiscoveryFeedBaseCardStore):
```

Comment:
Hmm, this still does not work:

```
Thread 1 "gjs" received signal SIGSEGV, Segmentation fault.
0x00007f749f60d6bc in gjs_typecheck_object (context=context@entry=0x1ef7400, object=..., object@entry=..., 
    expected_type=expected_type@entry=35040560, throw_error=throw_error@entry=true) at gi/object.cpp:2131
2131	    g_assert(priv->gtype == G_OBJECT_TYPE(priv->gobj));
(gdb) bt
#0  0x00007f749f60d6bc in gjs_typecheck_object(JSContext*, JS::HandleObject, GType, bool) (context=context@entry=0x1ef7400, object=..., object@entry=..., expected_type=expected_type@entry=35040560, throw_error=throw_error@entry=true) at gi/object.cpp:2131
#1  0x00007f749f5f2014 in gjs_value_to_g_argument(JSContext*, JS::HandleValue, GITypeInfo*, char const*, GjsArgumentType, GITransfer, bool, GArgument*) (context=context@entry=0x1ef7400, value=value@entry=..., type_info=type_info@entry=0x1f358f0, arg_name=arg_name@entry=0x0, arg_type=arg_type@entry=GJS_ARGUMENT_ARRAY_ELEMENT, transfer=GI_TRANSFER_NOTHING, may_be_null=false, arg=0x7fffcffebb60)
    at gi/arg.cpp:1683
#2  0x00007f749f5f4aed in gjs_array_to_explicit_array_internal(JSContext*, JS::HandleValue, GITypeInfo*, char const*, GjsArgumentType, GITransfer, bool, gpointer*, gsize*) (arr_p=<optimized out>, param_info=0x1f358f0, transfer=<optimized out>, length=<optimized out>, array_value=..., context=0x1ef7400) at gi/arg.cpp:919
#3  0x00007f749f5f4aed in gjs_array_to_explicit_array_internal(JSContext*, JS::HandleValue, GITypeInfo*, char const*, GjsArgumentType, GITransfer, bool, gpointer*, gsize*) (arr_p=<optimized out>, param_info=0x1f358f0, transfer=<optimized out>, length=<optimized out>, array_value=..., context=0x1ef7400) at gi/arg.cpp:1120
#4  0x00007f749f5f4aed in gjs_array_to_explicit_array_internal(JSContext*, JS::HandleValue, GITypeInfo*, char const*, GjsArgumentType, GITransfer, bool, gpointer*, gsize*) (context=context@entry=0x1ef7400, value=..., 
    value@entry=..., type_info=type_info@entry=0x7fffcffebe70, arg_name=arg_name@entry=0x7f74941043a4 "unordered_orderable_models", arg_type=arg_type@entry=GJS_ARGUMENT_ARGUMENT, transfer=transfer@entry=GI_TRANSFER_NOTHING, may_be_null=<optimized out>, contents=<optimized out>, length_p=<optimized out>) at gi/arg.cpp:1303
#5  0x00007f749f5f2194 in gjs_value_to_g_argument(JSContext*, JS::HandleValue, GITypeInfo*, char const*, GjsArgumentType, GITransfer, bool, GArgument*) (context=context@entry=0x1ef7400, value=..., type_info=type_info@entry=0x7fffcffebe70, arg_name=0x7f74941043a4 "unordered_orderable_models", arg_type=GJS_ARGUMENT_ARGUMENT, transfer=transfer@entry=GI_TRANSFER_NOTHING, may_be_null=false, arg=0x7fffcffebf60) at gi/arg.cpp:1918
#6  0x00007f749f5f5765 in gjs_value_to_arg(JSContext*, JS::HandleValue, GIArgInfo*, GIArgument*) (context=context@entry=0x1ef7400, value=..., arg_info=arg_info@entry=0x7fffcffec140, arg=arg@entry=0x7fffcffebf60) at gi/arg.cpp:2134
#7  0x00007f749f60324c in gjs_invoke_c_function(JSContext*, Function*, JS::HandleObject, JS::HandleValueArray const&, mozilla::Maybe<JS::MutableHandle<JS::Value> >, GIArgument*) (context=context@entry=0x1ef7400, function=function@entry=0x25d5800, obj=..., 
    obj@entry=..., args=..., js_rval=..., r_value=r_value@entry=0x0) at gi/function.cpp:973
#8  0x00007f749f604582 in function_call(JSContext*, unsigned int, JS::Value*) (context=0x1ef7400, js_argc=2, vp=0x2131b00)
    at gi/function.cpp:1353
```

I think even if we remove the `GDestroyNotify` from `g_task_return_pointer` this will still happen because `g_task_propagate_pointer` itself is `transfer full` (eg, it should transfer ownership through the return path). So the fact that it had ownership previously makes no difference.

As I mentioned before, I am pretty sure the problem is here: https://gitlab.gnome.org/GNOME/gjs/blob/master/gi/arg.cpp#L3457 . The array elements are released in both `transfer none` and `transfer full`, but not `transfer container`. At least for `transfer none` we take a ref on the container itself so the elements are not released.

I'll leave this as `transfer container` for now.

Reply:
Yeah, you're right... it'd probably be better to return a new reference from the finish method instead.

Reply:
> it'd probably be better to return a new reference from the finish method instead.

Hmm, not sure what you mean by this. So `transfer none` ?